Enable WinForms designer to open

davkean opened the pull request:
Enable WinForms designer to open

Fixes: #3823, #4804

This ports WinFormEditorProvider from CPS, and changes behavior from it and legacy:

- Users can now control (using Open With... -> Set As Default) whether the designer opens by default for Forms, User Controls and Components separately from each other. Legacy/CPS have a single setting that controls this. This is helpful because the Component designer is less useful than the Form designer, so you might want the latter to open but not the former.

- This new setting roams and is shared across all languages. Legacy/CPS is per language per VS install.

- Component Designer is turned off by default. Users can open it via the View Designer command, or use above Open With to set it to open by default.

- The name of the designer editor has changed from "CSharp/Visual Basic Form Editor" for Forms, User Controls and Components in legacy to:

Type|Name
---|---
Form|Form (Windows Forms) Designer 
Component|Component (Windows Forms) Designer
UserControl|User Control (Windows Forms) Designer

I feel these names are better descriptions of what they open and are consistent with the "XAML Designer".

This is a draft PR because we're blocked on CPS producing a package with an updated `IProjectSpecificEditorProvider` implementation. tag @adrianvmsft.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/WindowsForms/WindowsFormsEditorProvider.EditorInfo.cs:
```diff
@@ -0,0 +1,33 @@
+﻿/********************************************************
```
Will fix this.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/WindowsForms/WindowsFormsEditorProvider.cs:
```diff
@@ -0,0 +1,138 @@
+    ///     2) Persists whether the designer opens by default when the user uses Open With -> Set As Default.
+    /// </summary>
+    [Export(typeof(IProjectSpecificEditorProvider))]
+    [AppliesTo(ProjectCapability.DotNet)]
```
Unsure yet what capability to tie this lightup to.

Tanya-Solyanik replied:
`DotNet` in not in this list - https://github.com/microsoft/VSProjectSystem/blob/master/doc/overview/project_capabilities.md, what projects have it, both core and classic?

davkean replied:
Classic does not have .NET, but it doesn't matter this code never runs in the case of a legacy project system as it is CPS specific.

Tanya-Solyanik replied:
By default .NET classic projects use the legacy project system, however we can create a CPS-based project  manually and it works fine. So this will disable double-click in these projects and we say that this is OK because they are not supported officially and we'll face it later if needed?

davkean commented:
tag @Tanya-Solyanik @lifengl @DustinCampbell

drewnoakes commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/ProjectSystemOptions.cs:
```diff
@@ -32,25 +33,44 @@ public ProjectSystemOptions(IEnvironmentHelper environment, IVsUIService<SVsSett
             await _joinableTaskContext.Factory.SwitchToMainThreadAsync(cancellationToken);
-            return _settingsManager.Value?.GetValueOrDefault(FastUpToDateEnabledSettingKey, true) ?? true;
+
+            return _settingsManager.Value.GetValueOrDefault(name, defaultValue);
```
Is there no way `_settingsManager.Value` can be `null`? (Also below.)

davkean replied:
There's a theoretical case where GetService can return null during IDE shutdown. I don't think we want to handle that case.

drewnoakes commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/WindowsForms/WindowsFormsEditorProvider.EditorInfo.cs:
```diff
@@ -0,0 +1,30 @@
+
+using System;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.WindowsForms
```
Consider `#nullable enable` to all new files.

drewnoakes commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/IProjectSystemOptions.cs:
```diff
@@ -44,5 +44,15 @@ bool IsProjectOutputPaneEnabled
+        /// <summary>
+        ///     Gets a value indicating whether the designer is the default editor for the specified designer category.
+        /// </summary>
+        Task<bool> GetUseDesignerByDefault(string designerCategory, bool defaultValue, CancellationToken cancellationToken = default);
```
`Async` suffix on this and below (mirroring implementation).

merriemcgaw commented:
tagging @dreddy-work and @Shyam-Gupta to take a look too.

Shyam-Gupta commented:
@davkean Will there be a project system tools option that user can enable to have WinForms designer open by default when user selects a component file ?

Tanya-Solyanik commented:
@davkean  - do I understand correctly that these changes go into project system that runs core applications only? I'm concerned about the effect of this change on the classic .NET framework applications. If the user disables the default designer for the Core Form (this project is opened by CPS), what would happen when the same user double-clicks on a .NET classic Framework Form that uses the legacy project system?

Tanya-Solyanik commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/VSResources.resx:
```diff
@@ -218,4 +218,13 @@ In order to debug this project, add an executable project to this solution which
   <data name="Restore_EmptyTargetFrameworkMoniker" xml:space="preserve">
     <value>The value of the 'TargetFrameworkMoniker' and 'NuGetTargetMoniker' properties in the '{0}' configuration are both empty. This configuration will not contribute to NuGet restore, which may result in restore and build errors.</value>
   </data>
+  <data name="WindowsFormEditor_DisplayName" xml:space="preserve">
```
pardon my ignorance, but why are the old names not removed?

davkean replied:
Do you mean from legacy?

Tanya-Solyanik replied:
yes, are they still used?

Tanya-Solyanik replied:
Did you have "old" names in CPS before your change?

davkean replied:
CPS lives in a different code base. That will be deleted from there.

davkean commented:
@Tanya-Solyanik This does not affect legacy at all, just CPS.

davkean commented:
This PR has been updated, changes:

- I've slightly changed the description of the "designers" to match the exact format of the rest of the designers/editors.
- Introduced the ability for a feature flag to override whether the designer opens for .NET Core 3.0 projects. @merriemcgaw @Tanya-Solyanik @DustinCampbell The intention is that you'll add this feature flag to your standalone VSIX which will "light up" for .NET Core 3.0 and above.

drewnoakes commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/WindowsForms/WindowsFormsEditorProvider.cs:
```diff
@@ -0,0 +1,176 @@
+            // Project properties are more expensive than above, so only
+            // read them if absolutely must open the designer for the file
+            if (!await IsDesignerEnabledAsync())
+                return null;
```
This method serialises three async operations. Would it reduce delay when opening a file if they were all kicked off at the same time before being awaited? Based on your comment you may wish to do this only for the first two.

drewnoakes commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/PropertyExtensions.cs:
```diff
@@ -30,5 +30,27 @@ public static async Task<Guid> GetValueAsGuidAsync(this IProperty property)
+        ///     <paramref name="property"/> is <see langword="null"/>.
+        /// </exception>
+        public static async Task<TEnum> GetValueAsEnumAsync<TEnum>(this IProperty property)
+            where TEnum : struct
```
You can use an enum type constraint here.

davkean replied:
Tried that first, unfortunately the parse isn't `T : enum`

drewnoakes replied:
I think it should work:

https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQAcoKgW1QOzgBsiBjACwlIGsBLfAcwBoYRaimATEAagB8AAgCYAjAFgAUAIDMAAmGyAwrIDek2RtkRCOWQFFVsgL7rNM+QBZZAWQA8AFQB8ACgCUsgO6Uks+7JCyyDAIcKQwTPo6qqaamno6AHT2CACeAApYyBDOAEQAgjkRtAz4APZIilBZAcFwEBGlcDC+skjIxDCuANwxxpJGQA==

davkean replied:
Ah, didn't think to put struct in there as well. Thanks!

drewnoakes replied:
Yeah it seems redundant.

davidwengier commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/ConfigurationGeneral.xaml:
```diff
@@ -52,6 +52,9 @@
   <StringProperty Name="Configuration"
                   Visible="false" />
 
+  <StringProperty Name="WindowsFormsDesignerMode"
```
Why not define this as EnumProperty with the values here, and generated enum, rather than handle the mapping yourself and creating the enum yourself?

davkean replied:
Does it actually generate a enum? They were still strings.

davidwengier replied:
It should generate an enum (called WindowsFormsDesignerModeValues I guess) in the .cs file in the obj directory. From memory the values still get read as strings, but the cps property providers take care of checking for valid values, so consuming code can just use `string.Equals` for a cheaper comparison. It's possible that last statement is wrong however, and that's just the pattern of how CPS does it.

davkean replied:
i'll look into it, it actually looks a lot more complicated than just reading a string value and parsing.

davkean commented:
Heads up, I dialed this PR back to everything before we turned it off for .NET Core via feature which we'll merge in a follow up. I've made a couple of minor changes since last update, and added tests. I'll merge this in and do a follow up for the feature flag.
